src/zenml/cli/stack_recipes.py

```diff
@@ -0,0 +1,1013 @@
+                "https://learn.hashicorp.com/tutorials/terraform/install-cli "
+                "to install it."
+            )
+
+    def _is_installed(self) -> bool:
+        """Checks if terraform is installed on the host system.
+
+        Returns:
+            True if terraform is installed, false otherwise.
+        """
+        # check terraform version to verify installation.
+        ret_code, _, _ = self.tf.cmd("-version")
+        return bool(ret_code == 0)
+
+    def apply(self) -> str:
+        """Function to call terraform init and terraform apply.
+
+        The init call is not repeated if any successful execution has
+        happened already, to save time.
+
+        Returns:
+            The path to the stack YAML configuration file as a string.
+
+        Raises:
+            RuntimeError: if terraform init fails.
+        """
+        # this directory gets created after a successful init
+        previous_run_dir = os.path.join(self.tf.working_dir, ".ignoreme")
+        if fileio.exists(previous_run_dir):
+            logger.info(
+                "Terraform already initialized, "
+                "terraform init will not be executed."
+            )
+        else:
+            ret_code, _, _ = self.tf.init(capture_output=False)
+            if ret_code != 0:
+                raise RuntimeError("The command 'terraform init' failed.")
+            fileio.mkdir(previous_run_dir)
+
+        # get variables from the recipe as a python dictionary
+        vars = self._get_vars(self.tf.working_dir)
+
+        # once init is successful, call terraform apply
+        self.tf.apply(
+            var=vars,
+            input=False,
+            capture_output=False,
+            raise_on_error=True,
+        )
+
+        # return the path of the stack yaml file
+        _, stack_file_path, _ = self.tf.output("stack-yaml-path")
+        return str(stack_file_path)
+
+    def _get_vars(self, path: str) -> Any:
+        """Get variables as a dictionary from values.tfvars.json.
+
+        Args:
+            path: the path to the stack recipe.
+
+        Returns:
+            A dictionary of variables to use for the stack recipes
+            derived from the tfvars.json file.
+
+        Raises:
+            FileNotFoundError: if the values.tfvars.json file is not
+                found in the stack recipe.
+        """
+        import json
+
+        variables_file_path = os.path.join(path, VARIABLES_FILE)
+        if not fileio.exists(variables_file_path):
+            raise FileNotFoundError(
+                "The file values.tfvars.json was not found in the "
+                f"recipe's directory at {variables_file_path}. Please "
+                "verify if it exists."
+            )
+
+        # read values into a dict and return
+        with fileio.open(variables_file_path, "r") as f:
+            variables = json.load(f)
+            f.close()
+            return variables
```

Comment:
```suggestion
        with fileio.open(variables_file_path, "r") as f:
            variables = json.load(f)
        return variables
```
The `with` statement already takes care of closing the file for us here.

Reply:
okay, thanks for pointing out!